.github/actions/smart-ci/smart_ci.py

```diff
@@ -0,0 +1,180 @@
+
+        invalid_components = self.all_defined_components.difference(all_possible_components)
+        if invalid_components:
+            error_msg = f"components are invalid: " \
+                        f"{invalid_components} are not listed in labeler config: {all_possible_components}"
+            raise jsonschema.exceptions.ValidationError(error_msg)
+
+        for component_name, data in self.config.items():
+            dependent_components = set(data.get('dependent_components', dict()).keys()) if data else set()
+
+            invalid_dependents = dependent_components.difference(all_possible_components)
+            if invalid_dependents:
+                error_msg = f"dependent_components of {component_name} are invalid: " \
+                            f"{invalid_dependents} are not listed in components config: {all_possible_components}"
+                raise jsonschema.exceptions.ValidationError(error_msg)
+
+    def get_affected_components(self, changed_components_names: set) -> dict:
+        """Returns changed components, their dependencies and validation scope for them"""
+        affected_components = dict()
+
+        # If some changed components were not defined in config or no changed components detected at all,
+        # run full scope for everything (just in case)
+        changed_not_defined_components = changed_components_names.difference(self.all_defined_components)
+        if not changed_components_names or changed_not_defined_components:
+            self.log.info(f"Changed components {changed_not_defined_components} are not defined in smart ci config, "
+                          "run full scope")
+            affected_components.update({name: self.FullScope for name in self.all_possible_components})
+            return affected_components
+
+        # Else check changed components' dependencies and add them to affected
+        for name in changed_components_names:
+            component_data = self.config.get(name, dict())
+            dependent_components = component_data.get('dependent_components', dict()) if component_data else \
+                {name: self.FullScope for name in self.all_possible_components}
+
+            for dep_name, scope in dependent_components.items():
+                _scope = affected_components.get(dep_name, set()).union(scope)
+                affected_components.update({dep_name: _scope if scope else self.FullScope})
+
+        # If the component was explicitly changed, run full scope for it
+        affected_components.update({name: self.FullScope for name in changed_components_names})
+        self.log.info(f"Changed components with dependencies: {affected_components}")
+
+        # For non-affected components that are not defined in config - run full scope
+        affected_components.update({name: self.FullScope for name in self.all_possible_components
+                                    if name not in self.all_defined_components})
```

Comment:
> Looks like if component is not in list of smart ci config, it cannot be used in smart CI if conditions.
The problem is - when you refer to an action output or to an environment variable in GHA, you can write anything here (e.g. outputs.random_non_existing_var or env.random_non_existing_var), and it will **not** fail an will be evaluated as None, which is evaluated as "false" in "if" conditions - that's the way it works and we cannot change that. 

That's why I've added this line here, plus locally I experimented with a basic regexp-based linter for GHA pipeline to track and reveal typos in pipelines (I'm thinking of adding it directly to smart-ci action as a pre-step)